packages/caravan-health/src/privacy.ts

```diff
@@ -0,0 +1,285 @@
+  client: BlockchainClient,
+): number {
+  const numberOfInputs: number = transaction.vin.length;
+  const numberOfOutputs: number = transaction.vout.length;
+
+  const spendType: SpendType = determineSpendType(
+    numberOfInputs,
+    numberOfOutputs,
+  );
+  const score: number = SpendTypeScores(
+    spendType,
+    numberOfInputs,
+    numberOfOutputs,
+  );
+  if (
+    isSelfPayment(transaction, client) &&
+    spendType !== SpendType.Consolidation
+  ) {
+    return score * DENIABILITY_FACTOR;
+  }
+  return score;
+}
+
+/*
+Determine whether a Bitcoin transaction is a self-payment, meaning the sender and recipient
+are the same entity. To check this, we make an RPC call to the watcher wallet asking for the
+amount that a given address holds. If the call returns a number then it is part of wallet
+otherwise it is not a self payment.
+*/
+function isSelfPayment(
+  transaction: Transaction,
+  client: BlockchainClient,
+): boolean {
+  transaction.vout.forEach(async (op) => {
+    if ((await client.getAddressStatus(op.scriptPubkeyAddress)) === undefined) {
+      return false;
+    }
+  });
+  return true;
+}
+
+/* 
+In order to score for address reuse we can check the amount being hold by reused UTXOs 
+with respect to the total amount
+*/
+export async function addressReuseFactor(
+  utxos: Array<WalletUTXOs>,
+  client: BlockchainClient,
+): Promise<number> {
+  let reusedAmount: number = 0;
+  let totalAmount: number = 0;
+
+  for (const utxo of utxos) {
+    const address = utxo.address;
+    const reused = await isReusedAddress(address, client);
+
+    const amount = utxo.utxos.reduce((sum, utxo) => sum + utxo.value, 0);
+    if (reused) {
+      reusedAmount += amount;
+    }
+    totalAmount += amount;
+  }
+
+  return reusedAmount / totalAmount;
+}
+
+async function isReusedAddress(
+  address: string,
+  client: BlockchainClient,
+): Promise<boolean> {
+  let txs: Array<Transaction> = await client.getAddressTransactions(address);
+  if (txs.length > 1) {
+    return true;
+  }
+  return false;
+}
+/*
+If we are making payment to other wallet types then the privacy score should decrease because 
+the change received will be to an address type matching our wallet and it will lead to a deduction that
+we still own that amount.
+*/
+export function addressTypeFactor(
+  transactions: Array<Transaction>,
+  walletAddressType: string,
+): number {
+  let P2WSH: number = 0;
+  let P2PKH: number = 0;
+  let P2SH: number = 0;
+  let atf: number = 1;
+  transactions.forEach((tx) => {
+    tx.vout.forEach((output) => {
+      let address = output.scriptPubkeyAddress;
+      if (address.startsWith("bc")) {
+        //Bech 32 Native Segwit (P2WSH) or Taproot
+        P2WSH += 1;
+      } else if (address.startsWith("1")) {
+        // Legacy (P2PKH)
+        P2PKH += 1;
+      } else {
+        // Segwith (P2SH)
+        P2SH += 1;
+      }
+    });
+  });
+
+  if (walletAddressType == "P2WSH" && P2WSH != 0 && (P2SH != 0 || P2PKH != 0)) {
```

Comment:
I think a switch might be cleaner here and then can also switch on a match with the exported type from @caravan/bitcoin